Code/GraphMol/SynthonSpaceSearch/SynthonSpace.cpp

```diff
@@ -0,0 +1,719 @@
+          [&](const size_t prevVal, const SynthonSpaceHitSet &hs) -> size_t {
+            return prevVal + hs.numHits;
+          });
+      allHits.insert(allHits.end(), theseHits.begin(), theseHits.end());
+    }
+  }
+
+  if (params.buildHits) {
+    std::sort(allHits.begin(), allHits.end(),
+              [&](const SynthonSpaceHitSet &hs1,
+                  const SynthonSpaceHitSet &hs2) -> bool {
+                if (hs1.reactionId == hs2.reactionId) {
+                  return hs1.numHits < hs2.numHits;
+                } else {
+                  return hs1.reactionId < hs2.reactionId;
+                }
+              });
+    // Keep track of the result names so we can weed out duplicates by
+    // reaction and synthons.  Different splits may give rise to the same
+    // synthon combination.  This will keep the same molecule produced via
+    // different reactions which I think makes sense.  The resultsNames will
+    // be accumulated even if the molecule itself doesn't make it into the
+    // results set, for example if it isn't a random selection or it's
+    // outside maxHits or hitStart.
+    std::set<std::string> resultsNames;
+    // The random sampling, if requested, doesn't always produce the required
+    // number of hits in 1 sweep.  Also, totHits, being an upper bound, may
+    // not be achievable.
+    size_t tmpMaxHits(params.maxHits);
+    int numSweeps = 0;
+    while (results.size() < std::min(tmpMaxHits, totHits) && numSweeps < 10) {
+      buildHits(allHits, query, params, totHits, resultsNames, results);
+      // totHits is an upper bound, so may not be reached.
+      if (params.maxHits == -1 || !params.randomSample) {
+        break;
+      }
+      numSweeps++;
+    }
+  }
+  return SubstructureResults{std::move(results), totHits};
+}
+
+namespace {
+inline std::vector<std::string> splitLine(const std::string &str,
+                                          const std::regex &regexz) {
+  return {std::sregex_token_iterator(str.begin(), str.end(), regexz, -1),
+          std::sregex_token_iterator()};
+}
+
+// The Enamine/Chemspace files come with the connection points on the
+// synthons marked with [U], [Np], [Pu], [Am].  These need to be converted
+// to dummy atoms with isotope labels (1, 2, 3, 4 respectively) which can
+// be done safely on the SMILES string.
+void fixConnectors(std::string &smiles) {
+  const static std::vector<std::regex> connRegexs{
+      std::regex(R"(\[U\])"), std::regex(R"(\[Np\])"), std::regex(R"(\[Pu\])"),
+      std::regex(R"(\[Am\])")};
+  const static std::vector<std::string> repls{"[1*]", "[2*]", "[3*]", "[4*]"};
+  for (size_t i = 0; i < connRegexs.size(); ++i) {
+    smiles = std::regex_replace(smiles, connRegexs[i], repls[i]);
+  }
+}
+
+// Return a bitset giving the different connector types in this
+// molecule.
+boost::dynamic_bitset<> getConnectorPattern(
+    const std::vector<std::unique_ptr<ROMol>> &fragSet) {
+  boost::dynamic_bitset<> conns(4);
```

Comment:
It's probably worth making the max number of connections a constexpr at global scope (and maybe in the header)